[Skia] Fix support for DragImage with GTK

TingPing opened the pull request:
#### 73b85d6c9c77de1051c0521be9c8686202a14643
<pre>
[Skia] Fix support for DragImage with GTK
<a href="https://bugs.webkit.org/show_bug.cgi?id=271871">https://bugs.webkit.org/show_bug.cgi?id=271871</a>

Reviewed by Carlos Garcia Campos.

* Source/WebCore/platform/gtk/DragImageGtk.cpp:
(WebCore::dragImageSize):
(WebCore::scaleDragImage):
(WebCore::dissolveDragImageToFraction):
* Source/WebKit/UIProcess/API/gtk/DragSourceGtk3.cpp:
(WebKit::DragSource::begin):
* Source/WebKit/WebProcess/WebCoreSupport/gtk/WebDragClientGtk.cpp:
(WebKit::convertSkiaImageToShareableBitmap):
(WebKit::WebDragClient::startDrag):

Canonical link: <a href="https://commits.webkit.org/276871@main">https://commits.webkit.org/276871@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5390a28f5acc001bd05aaa0eda50ec2d819347b3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46037 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25166 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48622 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48708 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42077 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48344 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29480 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22591 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37644 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46615 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22254 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39717 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18834 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19650 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-transforms/transform3d-preserve3d-001.html, imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002.html (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40818 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4081 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42347 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41164 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50496 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21033 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17537 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44804 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22333 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43698 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22692 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6408 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22027 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/gtk/DragImageGtk.cpp:
```diff
@@ -21,13 +21,21 @@
+#include <skia/core/SkCanvas.h>
+#include <skia/core/SkImageInfo.h>
+#include <skia/core/SkPixmap.h>
+IGNORE_CLANG_WARNINGS_END
```
I think the warnings only need to be ignored for SkPixmap.h.

carlosgcampos commented on Source/WebCore/platform/gtk/DragImageGtk.cpp:
```diff
@@ -73,8 +81,19 @@ DragImageRef scaleDragImage(DragImageRef image, FloatSize scale)
+    if (!bitmap.peekPixels(&pixmap))
+        return nullptr;
+
+    SkSamplingOptions options; // FIXME: There may be a highest quality scaling option.
```
This is a small image, so we can probably use a high quality interpolation, use `SkSamplingOptions(SkCubicResampler::CatmullRom())`.

carlosgcampos commented on Source/WebCore/platform/gtk/DragImageGtk.cpp:
```diff
@@ -93,11 +112,26 @@ DragImageRef dissolveDragImageToFraction(DragImageRef image, float fraction)
+    paint.setAlphaf(fraction);
+
+    SkSamplingOptions options;
+    canvas->drawImage(image, 0, 0, options, &paint);
```
I think you can just pass `{ }` as sampling options.

carlosgcampos commented on Source/WebKit/WebProcess/WebCoreSupport/gtk/WebDragClientGtk.cpp:
```diff
@@ -65,16 +64,39 @@ static RefPtr<ShareableBitmap> convertCairoSurfaceToShareableBitmap(cairo_surfac
+
+    ASSERT(graphicsContext->hasPlatformContext());
+    auto* canvas = graphicsContext->platformContext();
+    canvas->drawImage(image, 0, 0);
```
This could be just `graphicsContext->platformContext()->drawImage(image, 0, 0);`

carlosgcampos commented on Source/WebKit/WebProcess/WebCoreSupport/gtk/WebDragClientGtk.cpp:
```diff
@@ -65,16 +64,39 @@ static RefPtr<ShareableBitmap> convertCairoSurfaceToShareableBitmap(cairo_surfac
-    auto& dragImage = dragItem.image;
-    auto bitmap = convertCairoSurfaceToShareableBitmap(dragImage.get().get());
     std::optional<ShareableBitmap::Handle> handle;
+    auto dragSurface = dragItem.image.get().get();
```
```suggestion
    auto* dragSurface = dragItem.image.get().get();
```

carlosgcampos reviewed (approved):
Thanks!

TingPing committed:
[Skia] Fix support for DragImage with GTK
https://bugs.webkit.org/show_bug.cgi?id=271871

Reviewed by Carlos Garcia Campos.

* Source/WebCore/platform/gtk/DragImageGtk.cpp:
(WebCore::dragImageSize):
(WebCore::scaleDragImage):
(WebCore::dissolveDragImageToFraction):
* Source/WebKit/UIProcess/API/gtk/DragSourceGtk3.cpp:
(WebKit::DragSource::begin):
* Source/WebKit/WebProcess/WebCoreSupport/gtk/WebDragClientGtk.cpp:
(WebKit::convertSkiaImageToShareableBitmap):
(WebKit::WebDragClient::startDrag):

Canonical link: https://commits.webkit.org/276871@main
